### DISCUSSION TITLE
Pitfalls of using `std::sync::Mutex` with the multithreaded runtime

### DISCUSSION THREAD

**Discussion by samueltardieu** [tokio-rs/tokio#6843] [General]
The documentation of `tokio::sync::Mutex` contains:

> Contrary to popular belief, it is okay and often preferred to use the ordinary [Mutex](https://doc.rust-lang.org/nightly/std/sync/mutex/struct.Mutex.html) from the standard library in asynchronous code.

This includes a well-articulated explanation of why this is the case. I concur with this explanation as long as the mutex is used within a single-threaded asynchronous runtime.

However, using `std::sync::Mutex` with a multithreaded executor may cause another thread to be temporarily blocked when the mutex is held. If this thread is running an asynchronous executor, it will be unable to execute other asynchronous tasks during this time, and other threads may not be available to execute these tasks through work stealing.

If asynchronous tasks accessing this mutex are dispersed across the executor threads, there is a risk that multiple executor threads could become blocked. Shouldn't this be highlighted?

**Comment by Darksonn** [comment]
Ultimately, it's perfectly fine to use `std::sync::Mutex` from a multi-threaded runtime. In fact, lots of things in `tokio::sync` work that way internally. As long as you don't hold the mutex for very long, there's no problem.
